channeldb/migration13/migration_test.go

```diff
@@ -0,0 +1,88 @@
+package migration13
+
+import (
+	"testing"
+
+	"github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/channeldb/migtest"
+)
+
+var (
+	// pre is the data in the payments root bucket in database version 12 format.
+	pre = map[string]interface{}{
+		// A failed payment.
+		"0x02acee76ebd53d00824410cf6adecad4f50334dac702bd5a2d3ba01b91709f0e": map[string]interface{}{
+			"payment-attempt-info":  "0x00000000000000012997a72e129fc9d638ef2fa4e233567d808d4f18a4f087637582427962eb3bf800005ce600000000004c4b402102ec12e83eafe27ce6d03bbe0c0de4b79fe2b9934615c8aa7693f73d2e41b089700000000121028c2dd128c7a6c1a0fceb3e3eb5ed55e0a0ae1a939eb786b097322d830d47db75005ca4000001000000005ce600000000004c4b400000000000",
+			"payment-creation-info": "0x02acee76ebd53d00824410cf6adecad4f50334dac702bd5a2d3ba01b91709f0e00000000004c4b40000000005e4fb7ab00000000",
+			"payment-fail-info":     "0x03",
+			"payment-sequence-key":  "0x0000000000000001",
+		},
+
+		// A settled payment.
+		"0x62eb3f0a48f954e495d0c14ac63df04a67cefa59dafdbcd3d5046d1f5647840c": map[string]interface{}{
```

Comment:
Yes, I produced them using the helper function `DumpDb` on a prior version database. These assertions are indeed low-level, but imo that is okay because the migration itself is also low-level. It doesn't use the usual set of (duplicated) serialize/deserialize functions. The migration modifies byte strings directly and with that knowledge, it isn't difficult to verify the correctness of the assertions. That makes it more than just locking the code at a specific point.

I think the underlying question is whether we want to stick with copying two sets of serialize/deserialize functions as we did previously, or go for this low-level approach. Migration 12 already started on the path of more low-level migration and testing and I continued this here. @halseth @cfromknecht thoughts?

Reply:
i agree that the test as is seems fragile. just because the bytes are correct doesn't mean that the deserialization is actually reconstructing the proper in-memory object after decoding. 

I think it depends how these were ultimately generated. in #3693, we initially started with all the functions copied in so that we could do the in-memory assertions in case other parts of the pr changed. only once all changes stopped did we remove the serialization methods and end with just storing, getting, and comparing the raw byte slices generated by the final variant. there's really no other way to know that our other modification haven't invalidated the deserialization, which happened on a few occasions during #3693

Reply:
Tried to do an old-style migration test, using serialize v12 functions to populate the db and deserialize v13 to assert, but it got totally out of hand. A lot of code is required and also modifications need to be made to get the code to work outside of its usual context. Those modifications are a risk themselves.

I think the low-level migration test is still the better way to go, everything taken into account. I've introduced sets of constants to make them more readable.